# HW3

In [29]:
import numpy as np
import pandas as pd
import tensorflow
import requests
import time
import yfinance as yf
from itertools import product
from sklearn.metrics import accuracy_score, classification_report
from sklearn.datasets import fetch_olivetti_faces
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.multiclass import OneVsRestClassifier, OneVsOneClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.svm import SVR
from sklearn.ensemble import AdaBoostRegressor
from lightgbm import LGBMRegressor
from xgboost import XGBRegressor
from sklearn.svm import SVC
from tabulate import tabulate
from sklearn.metrics import accuracy_score, classification_report
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

import xgboost as xgb
from lightgbm import LGBMClassifier

import warnings
warnings.filterwarnings("ignore", category=Warning)

## Atabey's Notes

- Q1: Nice clean implementation.
- Q2: Same as above.
- Q3: It works but the code has some problems. 
- Q4: You repeated the same code as above.

25 + 25 + 20 + 20

## Q1

For this question use the MNIST dataset (from tensorflow).

1. Write multi-class classification models using the following regular classification algorithms
   - Logistic regression
   - SVM
   - Decision Tree
   
   Employ one-vs-rest (OvR) and one-vs-one (OvO) in combining different classifiers.

2. Statistically analyze accuracy, precision, and recall for all of the models above in a table.

3. Write multi-class classification models using the following ensemble classification algorithms
   - Random forest
   - AdaBoost
   - XGBoost
   - LightGBM
   
   Employ one-vs-rest (OvR) and one-vs-one (OvO) in combining different classifiers.

4. Statistically analyze accuracy, precision, and recall for all of the models above in a table.

5. Compare your results from Parts 2 and 4.

First, let us load the data(there is no Nan values so no need imputation).Then define a function in order to do classification algorithms with ensemble learning strategies

In [3]:
(mx_train, my_train), (mx_test, my_test)=tensorflow.keras.datasets.mnist.load_data()

In [4]:
def classifier_model(X_train,X_test,y_train,y_test,strategy,model):
    res = strategy(model).fit(X_train, y_train)
    y_pred = res.predict(X_test)
    return classification_report(y_test, y_pred)

In [5]:
mx_train.shape

(60000, 28, 28)

As we can see data is in 3d form. So, we need to flatten it in order to use in a model:

In [6]:
mx_train_flat = mx_train.reshape(-1, 28*28)
mx_train_flat = mx_train_flat/255.0
mx_test_flat = mx_test.reshape(-1, 28*28)
mx_test_flat = mx_test_flat/255.0

In [7]:
mx_train_flat.shape

(60000, 784)

In [8]:
unique_values, counts = np.unique(my_train, return_counts=True)

# Print unique values and their counts
for value, count in zip(unique_values, counts):
    print(f"Value: {value}, Count: {count}")

Value: 0, Count: 5923
Value: 1, Count: 6742
Value: 2, Count: 5958
Value: 3, Count: 6131
Value: 4, Count: 5842
Value: 5, Count: 5421
Value: 6, Count: 5918
Value: 7, Count: 6265
Value: 8, Count: 5851
Value: 9, Count: 5949


Data is relatively balanced. \
Now let use do classification with classical models and strategies combinations :

In [9]:
print(classifier_model(mx_train_flat, mx_test_flat, my_train, my_test, OneVsRestClassifier,LogisticRegression(max_iter=4000)))

              precision    recall  f1-score   support

           0       0.95      0.98      0.96       980
           1       0.96      0.98      0.97      1135
           2       0.94      0.89      0.91      1032
           3       0.90      0.91      0.90      1010
           4       0.92      0.93      0.92       982
           5       0.89      0.86      0.88       892
           6       0.94      0.95      0.94       958
           7       0.94      0.93      0.93      1028
           8       0.87      0.87      0.87       974
           9       0.90      0.89      0.90      1009

    accuracy                           0.92     10000
   macro avg       0.92      0.92      0.92     10000
weighted avg       0.92      0.92      0.92     10000



In [10]:
print(classifier_model(mx_train_flat, mx_test_flat, my_train, my_test, OneVsOneClassifier,LogisticRegression(max_iter=4000)))

              precision    recall  f1-score   support

           0       0.97      0.98      0.98       980
           1       0.98      0.99      0.98      1135
           2       0.94      0.94      0.94      1032
           3       0.92      0.93      0.93      1010
           4       0.94      0.96      0.95       982
           5       0.92      0.90      0.91       892
           6       0.96      0.96      0.96       958
           7       0.95      0.94      0.95      1028
           8       0.92      0.91      0.91       974
           9       0.93      0.92      0.93      1009

    accuracy                           0.94     10000
   macro avg       0.94      0.94      0.94     10000
weighted avg       0.94      0.94      0.94     10000



In [11]:
print(classifier_model(mx_train_flat, mx_test_flat, my_train, my_test, OneVsRestClassifier,SVC()))

              precision    recall  f1-score   support

           0       0.98      0.99      0.98       980
           1       0.99      0.99      0.99      1135
           2       0.97      0.97      0.97      1032
           3       0.98      0.98      0.98      1010
           4       0.98      0.98      0.98       982
           5       0.99      0.98      0.99       892
           6       0.98      0.99      0.98       958
           7       0.98      0.97      0.97      1028
           8       0.97      0.98      0.98       974
           9       0.97      0.96      0.97      1009

    accuracy                           0.98     10000
   macro avg       0.98      0.98      0.98     10000
weighted avg       0.98      0.98      0.98     10000



In [12]:
print(classifier_model(mx_train_flat, mx_test_flat, my_train, my_test, OneVsOneClassifier,SVC()))

              precision    recall  f1-score   support

           0       0.98      0.99      0.99       980
           1       0.99      0.99      0.99      1135
           2       0.98      0.98      0.98      1032
           3       0.98      0.98      0.98      1010
           4       0.98      0.98      0.98       982
           5       0.99      0.98      0.98       892
           6       0.99      0.99      0.99       958
           7       0.98      0.97      0.97      1028
           8       0.97      0.98      0.97       974
           9       0.97      0.96      0.97      1009

    accuracy                           0.98     10000
   macro avg       0.98      0.98      0.98     10000
weighted avg       0.98      0.98      0.98     10000



In [13]:
print(classifier_model(mx_train_flat, mx_test_flat, my_train, my_test, OneVsRestClassifier,DecisionTreeClassifier(max_depth=6,criterion='entropy')))

              precision    recall  f1-score   support

           0       0.92      0.95      0.93       980
           1       0.92      0.96      0.94      1135
           2       0.86      0.83      0.85      1032
           3       0.84      0.82      0.83      1010
           4       0.88      0.84      0.86       982
           5       0.83      0.78      0.81       892
           6       0.94      0.89      0.92       958
           7       0.88      0.89      0.89      1028
           8       0.82      0.84      0.83       974
           9       0.80      0.87      0.83      1009

    accuracy                           0.87     10000
   macro avg       0.87      0.87      0.87     10000
weighted avg       0.87      0.87      0.87     10000



In [14]:
print(classifier_model(mx_train_flat, mx_test_flat, my_train, my_test, OneVsOneClassifier,DecisionTreeClassifier(max_depth=6,criterion='entropy')))

              precision    recall  f1-score   support

           0       0.97      0.97      0.97       980
           1       0.99      0.97      0.98      1135
           2       0.93      0.90      0.91      1032
           3       0.92      0.88      0.89      1010
           4       0.95      0.89      0.92       982
           5       0.92      0.86      0.89       892
           6       0.94      0.96      0.95       958
           7       0.94      0.93      0.93      1028
           8       0.84      0.92      0.88       974
           9       0.84      0.93      0.89      1009

    accuracy                           0.92     10000
   macro avg       0.92      0.92      0.92     10000
weighted avg       0.92      0.92      0.92     10000



Let us summarize the results(with weighted average):

| Model                | Strategy | Precision | Recall | F-1   | Accuracy |
|----------------------|----------|-----------|--------|-------|----------|
| Logistic Regression  | OnevsRest|    0.92    | 0.92  | 0.92 | 0.92    |
| Logistic Regression  | OnevsOne |    0.94    | 0.94  | 0.94 | 0.94    |
| SVM  | OnevsRest |    0.98    | 0.98  | 0.98 | 0.98    |
| SVM  | OnevsOne |    0.98    | 0.98  | 0.98 | 0.98    |
| Decision Tree  | OnevsRest |    0.87    | 0.87  | 0.87 | 0.87    |
| Decision Tree  | OnevsOne |    0.92    | 0.92  | 0.92 | 0.92    |



SVM takes a long time compared to others. While Decision Tree model generally has a bit lower result(tried different depths), overall results are fine. As the data is relatively balanced F-1 and other scores do not differ significantly. \
As a strategy, OneVsOne generally performs better(probably because the data is balanced) compared to OneVsRest strategy.

Now, let us use ensemble classification algorithms with different combinations of strategies :

In [15]:
models = {'Random Forest': RandomForestClassifier(n_estimators=10), 
          'AdaBoost': AdaBoostClassifier(algorithm='SAMME'),
          'XGBoost': xgb.XGBClassifier(objective='multi:softmax', num_class=10),
          'LightGBM': LGBMClassifier(num_leaves=10, n_estimators=50)}
strategies={'OneVsOne': OneVsOneClassifier,'OneVsRest': OneVsRestClassifier}

We will iterate over each combination and call the previously defined 'classifier_model' function, also measure the time

In [16]:
for model_name, model in models.items():
    for strategy_name, strategy in strategies.items():
        print(f"Training and evaluating {model_name} with {strategy_name} strategy...")
        start_time = time.time()
        print(classifier_model(mx_train_flat, mx_test_flat, my_train, my_test, strategy,model))
        end_time = time.time()
        print(f"Time taken: {end_time - start_time:.2f} seconds\n")

Training and evaluating Random Forest with OneVsOne strategy...


              precision    recall  f1-score   support

           0       0.97      0.99      0.98       980
           1       0.99      0.99      0.99      1135
           2       0.95      0.95      0.95      1032
           3       0.95      0.95      0.95      1010
           4       0.96      0.96      0.96       982
           5       0.96      0.95      0.95       892
           6       0.98      0.96      0.97       958
           7       0.97      0.95      0.96      1028
           8       0.92      0.95      0.94       974
           9       0.94      0.93      0.93      1009

    accuracy                           0.96     10000
   macro avg       0.96      0.96      0.96     10000
weighted avg       0.96      0.96      0.96     10000

Time taken: 39.70 seconds

Training and evaluating Random Forest with OneVsRest strategy...
              precision    recall  f1-score   support

           0       0.97      0.99      0.98       980
           1       0.99      0.98      0

| Model                | Strategy | Precision | Recall | F-1   | Accuracy | Time |
|----------------------|----------|-----------|--------|-------|----------|------|
| Random Forest  | OnevsOne |    0.96    | 0.96  | 0.96 | 0.96    | 55.11 |
| Random Forest  | OnevsRest |    0.95    | 0.95  | 0.95 | 0.95    | 45.40 |
| AdaBoost  | OnevsOne |    0.91    | 0.91  | 0.91 | 0.91    | 1226.32 |
| AdaBoost  | OnevsRest |    0.87    | 0.87  | 0.86 | 0.87    | 1531.53 |
| XGBoost  | OnevsOne |    0.98    | 0.98  | 0.98 | 0.98    | 2185.98 |
| XGBoost  | OnevsRest |    0.98    | 0.98  | 0.98 | 0.98    | 2649.34 |
| LightGBM  | OnevsOne |    0.96    | 0.96  | 0.96 | 0.96    | 85.91 |
| LightGBM  | OnevsRest |    0.95    | 0.95  | 0.95 | 0.95    | 61.47 |


XGBoost has highest score with highest time. Then Random Forest and LightGBM performs well with considerably short time. AdaBoost has comparably low performance and long time.

Both classical and ensemble algorithms work better with OneVsOne strategy.

## Q2

Repeat what you did in Q1 for Olivetti Faces dataset (from tensorflow).

In [28]:
olivetti = fetch_olivetti_faces()

# Access the data and target
X = olivetti.data
y = olivetti.target

By using stratify parameter,for each person 7 images of them will be in train,3 of them will be in a test set.

In [18]:
ox_train, ox_test, oy_train, oy_test = train_test_split(X, y, test_size=0.3,stratify=y) 

In [19]:
unique_values, counts = np.unique(oy_train, return_counts=True)

# Print the results
print("Unique values:", unique_values)
print("Counts:", counts)

Unique values: [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39]
Counts: [7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7 7
 7 7 7]


Again the data is balanced.

In [20]:
print(classifier_model(ox_train, ox_test, oy_train, oy_test, OneVsRestClassifier,LogisticRegression(max_iter=4000)))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00         3
           1       1.00      1.00      1.00         3
           2       1.00      0.67      0.80         3
           3       0.75      1.00      0.86         3
           4       1.00      1.00      1.00         3
           5       1.00      1.00      1.00         3
           6       1.00      1.00      1.00         3
           7       1.00      1.00      1.00         3
           8       1.00      1.00      1.00         3
           9       1.00      0.67      0.80         3
          10       1.00      1.00      1.00         3
          11       1.00      1.00      1.00         3
          12       0.67      0.67      0.67         3
          13       1.00      1.00      1.00         3
          14       1.00      1.00      1.00         3
          15       1.00      1.00      1.00         3
          16       1.00      1.00      1.00         3
          17       1.00    

In [21]:
print(classifier_model(ox_train, ox_test, oy_train, oy_test, OneVsOneClassifier,LogisticRegression(max_iter=4000)))

              precision    recall  f1-score   support

           0       0.75      1.00      0.86         3
           1       1.00      1.00      1.00         3
           2       1.00      1.00      1.00         3
           3       0.75      1.00      0.86         3
           4       1.00      0.67      0.80         3
           5       1.00      1.00      1.00         3
           6       1.00      1.00      1.00         3
           7       1.00      1.00      1.00         3
           8       1.00      1.00      1.00         3
           9       1.00      0.67      0.80         3
          10       1.00      1.00      1.00         3
          11       1.00      1.00      1.00         3
          12       1.00      0.67      0.80         3
          13       1.00      1.00      1.00         3
          14       1.00      1.00      1.00         3
          15       1.00      1.00      1.00         3
          16       1.00      1.00      1.00         3
          17       1.00    

In [22]:
print(classifier_model(ox_train, ox_test, oy_train, oy_test, OneVsRestClassifier,SVC(max_iter=1000)))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00         3
           1       0.75      1.00      0.86         3
           2       1.00      1.00      1.00         3
           3       1.00      0.67      0.80         3
           4       1.00      0.67      0.80         3
           5       0.75      1.00      0.86         3
           6       1.00      1.00      1.00         3
           7       1.00      1.00      1.00         3
           8       1.00      1.00      1.00         3
           9       1.00      1.00      1.00         3
          10       1.00      1.00      1.00         3
          11       1.00      1.00      1.00         3
          12       1.00      0.67      0.80         3
          13       1.00      1.00      1.00         3
          14       0.75      1.00      0.86         3
          15       1.00      1.00      1.00         3
          16       1.00      1.00      1.00         3
          17       0.75    

In [23]:
print(classifier_model(ox_train, ox_test, oy_train, oy_test, OneVsOneClassifier,SVC(max_iter=1000)))

              precision    recall  f1-score   support

           0       1.00      0.67      0.80         3
           1       1.00      1.00      1.00         3
           2       1.00      0.67      0.80         3
           3       0.67      0.67      0.67         3
           4       1.00      1.00      1.00         3
           5       1.00      1.00      1.00         3
           6       1.00      1.00      1.00         3
           7       1.00      1.00      1.00         3
           8       0.75      1.00      0.86         3
           9       1.00      0.67      0.80         3
          10       1.00      1.00      1.00         3
          11       1.00      0.67      0.80         3
          12       1.00      0.67      0.80         3
          13       1.00      1.00      1.00         3
          14       1.00      1.00      1.00         3
          15       1.00      1.00      1.00         3
          16       1.00      1.00      1.00         3
          17       1.00    

In [24]:
print(classifier_model(ox_train, ox_test, oy_train, oy_test, OneVsRestClassifier,DecisionTreeClassifier(max_depth=10,criterion='entropy')))

              precision    recall  f1-score   support

           0       1.00      0.33      0.50         3
           1       0.00      0.00      0.00         3
           2       1.00      0.33      0.50         3
           3       0.00      0.00      0.00         3
           4       0.50      0.67      0.57         3
           5       1.00      0.67      0.80         3
           6       1.00      1.00      1.00         3
           7       1.00      0.33      0.50         3
           8       0.75      1.00      0.86         3
           9       0.00      0.00      0.00         3
          10       0.00      0.00      0.00         3
          11       0.00      0.00      0.00         3
          12       0.00      0.00      0.00         3
          13       0.00      0.00      0.00         3
          14       0.33      0.33      0.33         3
          15       0.50      0.33      0.40         3
          16       1.00      0.33      0.50         3
          17       1.00    

In [25]:
print(classifier_model(ox_train, ox_test, oy_train, oy_test, OneVsOneClassifier,DecisionTreeClassifier(max_depth=10,criterion='entropy')))

              precision    recall  f1-score   support

           0       0.75      1.00      0.86         3
           1       1.00      0.67      0.80         3
           2       1.00      0.67      0.80         3
           3       0.50      0.67      0.57         3
           4       1.00      0.33      0.50         3
           5       1.00      1.00      1.00         3
           6       1.00      0.33      0.50         3
           7       1.00      1.00      1.00         3
           8       0.75      1.00      0.86         3
           9       1.00      0.67      0.80         3
          10       1.00      1.00      1.00         3
          11       1.00      0.67      0.80         3
          12       0.67      0.67      0.67         3
          13       0.67      0.67      0.67         3
          14       1.00      1.00      1.00         3
          15       0.60      1.00      0.75         3
          16       1.00      1.00      1.00         3
          17       1.00    

| Model                | Strategy | Precision | Recall | F-1   | Accuracy |
|----------------------|----------|-----------|--------|-------|----------|
| Logistic Regression  | OnevsRest|    0.97    | 0.97  | 0.96 | 0.97    |
| Logistic Regression  | OnevsOne |    0.98    | 0.97  | 0.97 | 0.97    |
| SVM  | OnevsRest |    0.96    | 0.94  | 0.94 | 0.94    |
| SVM  | OnevsOne |    0.98    | 0.97  | 0.98 | 0.97    |
| Decision Tree  | OnevsRest |    0.65    | 0.50  | 0.52 | 0.50    |
| Decision Tree  | OnevsOne |    0.90    | 0.83  | 0.84 | 0.83    |

Logistic Regression has an overall high score.Then SVM and lastly Decision Tree.(In Decision Tree, OneVsRest doesnt change that much even with different max_depths). Also OneVsOne strategy performs slightly better with all methods.

Now, let us use ensemble learning models:

In [26]:
models2 = {'Random Forest': RandomForestClassifier(n_estimators=10), 
          'AdaBoost': AdaBoostClassifier(algorithm='SAMME'),
          'XGBoost': xgb.XGBClassifier(objective='multi:softmax', num_class=40),
          'LightGBM': LGBMClassifier(num_leaves=40, n_estimators=50)}

At this part, i wanted to print all result at once(because of lightgbm outputs was a bit messy). First we will try all combinations and measure time and store them, then print:

In [27]:
res_q2=[]
for model_name, model in models2.items():
    for strategy_name, strategy in strategies.items():
        print(f"Training and evaluating {model_name} with {strategy_name} strategy...")
        start_time = time.time()
        modell = classifier_model(ox_train, ox_test, oy_train, oy_test, strategy,model)
        end_time = time.time()
        res_q2.append((modell,end_time - start_time))

Training and evaluating Random Forest with OneVsOne strategy...


Training and evaluating Random Forest with OneVsRest strategy...
Training and evaluating AdaBoost with OneVsOne strategy...
Training and evaluating AdaBoost with OneVsRest strategy...
Training and evaluating XGBoost with OneVsOne strategy...


KeyboardInterrupt: 

In [ ]:
i=0
for model_name, model in models2.items():
    for strategy_name, strategy in strategies.items():
        print(f"{model_name}--{strategy_name}\n")
        print(res_q2[i][0])
        print(f"Time : {res_q2[i][1]} \n  \n")
        i=i+1

Random Forest--OneVsOne

              precision    recall  f1-score   support

           0       0.75      1.00      0.86         3
           1       1.00      1.00      1.00         3
           2       1.00      1.00      1.00         3
           3       0.50      0.67      0.57         3
           4       0.33      0.33      0.33         3
           5       1.00      0.67      0.80         3
           6       0.75      1.00      0.86         3
           7       1.00      1.00      1.00         3
           8       0.50      0.67      0.57         3
           9       1.00      0.67      0.80         3
          10       1.00      1.00      1.00         3
          11       1.00      1.00      1.00         3
          12       0.50      0.33      0.40         3
          13       0.75      1.00      0.86         3
          14       1.00      1.00      1.00         3
          15       1.00      0.67      0.80         3
          16       1.00      1.00      1.00         3
  

| Model                | Strategy | Precision | Recall | F-1   | Accuracy | Time |
|----------------------|----------|-----------|--------|-------|----------|------|
| Random Forest  | OnevsOne |    0.90    | 0.88  | 0.88 | 0.88    | 48.33 |
| Random Forest  | OnevsRest |    0.76    | 0.76  | 0.73 | 0.76    | 6.58 |
| AdaBoost  | OnevsOne |    0.80    | 0.76  | 0.75 | 0.76    | 26.40 |
| AdaBoost  | OnevsRest |    0.89    | 0.86 | 0.86 | 0.86    | 494.05 |
| XGBoost  | OnevsOne |    0.00    | 0.03  | 0.00 | 0.03    | 4291.25 |
| XGBoost  | OnevsRest |    0.87    | 0.82  | 0.83 | 0.82    | 942.86 |
| LightGBM  | OnevsOne |    0.00    | 0.03  | 0.00 | 0.03    | 66.49 |
| LightGBM  | OnevsRest |    0.78    | 0.77  | 0.75 | 0.77    | 118.60 |


In ensemble learning models, while Random Forest tooks shortest time, it and Adaboost have the best performances. XGBoost takes extremely long time and the performance is not significantly high. LightGBM also ist almost same. \
When it comes to strategies, while Random Forest performs better with OneOvsOne, rest of the models do not. We know that the data is balanced but the size of the data is also small. For OneVsOne, there are 14 data to train which might be insufficient for a healthy analysis. \
In conclusion, if we consider overall performance with time, Random forest looks like a best choice.

If we compare the classical and ensemble models, classical models gives more trust with overall performance. For strategy part,OneOvsOne peforms better with classical models and in ensemble models just the opposite looks true.

## Q3

For this question use the World Bank's [Climate Change Knowledge Portal](https://climateknowledgeportal.worldbank.org/).

The Mediterranean Sea is bordered by the following 22 countries: Spain, France, Monaco, Italy, Slovenia, Croatia, Bosnia and Herzegovina, Montenegro, Albania, Greece, Turkey, Syria, Lebanon, Israel, Palestine, Egypt, Libya, Tunisia, Algeria, Morocco, Malta, and Cyprus. 

Using the World Bank's Climate Change Knowledge Portal, obtain the following data through their API:

- Monthly average surface temperatures
- Monthly average precipitation

Gather this time series from 1950 for all countries surrounding the Mediterranean Sea. 

1. Write a function that takes 3 parameters:

   - 'surface' vs 'precipitation'
   - a model type, and
   - a country name
   
   that returns a

   - regular OLS regression
   - decision tree regression
   - support vector regression
   - AdaBoost regression
   - LightGBM regression
   - XGBoost regression
   
   model against time for the given country.

3. Calculate every possible model combination for every country and evaluate each model.
4. Report the best model for surface temperatures and precipitation for each country in a table. 

(I used the codes from my past assignment in MAT381 for the data acquisition part)

In [30]:
countries = {
    'Spain': 'ESP',
    'France': 'FRA',
    'Monaco': 'MCO',
    'Italy': 'ITA',
    'Slovenia': 'SVN',
    'Croatia': 'HRV',
    'Bosnia and Herzegovina': 'BIH',
    'Montenegro': 'MNE',
    'Albania': 'ALB',
    'Greece': 'GRC',
    'Turkey': 'TUR',
    'Syria': 'SYR',
    'Lebanon': 'LBN',
    'Israel': 'ISR',
    'Palestine': 'PSE',
    'Egypt': 'EGY',
    'Libya': 'LBY',
    'Tunisia': 'TUN',
    'Algeria': 'DZA',
    'Morocco': 'MAR',
    'Malta': 'MLT',
    'Cyprus': 'CYP'
}
country_codes = ','.join(countries.values())
country_names = ','.join(countries.keys())

In [31]:
api_url = f'https://cckpapi.worldbank.org/cckp/v1/era5-x0.5_timeseries_pr,tas_timeseries_monthly_1950-2020_mean_historical_era5_era5_mean/{country_codes}?_format=json'

In [32]:
try:
    # Send a GET request to the API
    response = requests.get(api_url)

    # Check if the request was successful (status code 200)
    if response.status_code == 200:
        # The response content will contain the data you requested
        data = response.json()  # or response.text if the response is in plain text
    else:
        # Print an error message for non-200 status codes
        print(f"Request failed with status code {response.status_code}")
except Exception as e:
    # Handle any exceptions that may occur during the request
    print(f"An error occurred: {str(e)}")

In [33]:
df = pd.DataFrame(data)

Now, i defined a function below that takes data type, model type, and country name, then return the model which is constructed by using time and related data values. I thought seperating year and month might be better so i did. I also add the evaluation parts to the function(mae,mse, and r^2)

In [34]:
def wb_getmodels(data_type,model_type,country_name):
    if data_type == 'surface':
        data = pd.DataFrame(df['data']['tas']).reset_index()
    elif data_type == 'precipitation':
        data = pd.DataFrame(df['data']['pr']).reset_index()
    else:
        print('invalid data parameter.')

    dates = pd.to_datetime(data['index'])
    year = dates.dt.year
    month = dates.dt.month
    data2=pd.DataFrame({'year': year, 'month': month,'Target': data[country_name]})
    X=data2[['year','month']]
    y = data2['Target']

    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25)

    # OLS
    if model_type == 'OLS':
        model = LinearRegression()
        model.fit(X_train, y_train)
    
    # Decision Tree Regression
    elif model_type == 'Decision Tree':
        model = DecisionTreeRegressor()
        model.fit(X_train, y_train)
    
    # SVM
    elif model_type == 'SVR':
        model = SVR()
        model.fit(X_train, y_train)
    
    # AdaBoost Regression
    elif model_type == 'AdaBoost':
        model = AdaBoostRegressor()
        model.fit(X_train, y_train)
    
    # LightGBM Regression
    elif model_type == 'LightGBM':
        model = LGBMRegressor()
        model.fit(X_train, y_train)
    
    # XGBoost Regression
    elif model_type == 'XGBoost':
        model = XGBRegressor()
        model.fit(X_train, y_train)
    else:
        print('Invalid model type')
    
    y_pred = model.predict(X_test)
    mae = mean_absolute_error(y_test, y_pred)
    mse = mean_squared_error(y_test, y_pred)
    r2 = r2_score(y_test, y_pred)
    
    return mae,mse,r2


Here are the lists of all countries and models:

In [35]:
country_list=countries.values()
model_lists=['OLS','Decision Tree','SVR','AdaBoost','LightGBM','XGBoost']

There is a result dataframe which has the country, data type, model and scores. I iterate each data to each combination of country and model,then store the results in this dataframe:

In [36]:
results_df = pd.DataFrame(columns=['Country','Data', 'Model', 'MAE', 'MSE', 'R2'])

# Iterate over countries and models
for country in country_list:
    for model in model_lists:
        # Call the function
        mae, mse, r2 = wb_getmodels('surface',model,country)
        # Append results to DataFrame
        results_df = results_df.append({'Country': country, 'Data': 'Avg surf. temp.','Model': model, 'MAE': mae, 'MSE': mse, 'R2': r2}, ignore_index=True)

[LightGBM] [Warning] Auto-choosing col-wise multi-threading, the overhead of testing was 0.000070 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 85
[LightGBM] [Info] Number of data points in the train set: 639, number of used features: 2
[LightGBM] [Info] Start training from score 13.579296
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, 

In [ ]:
for country in country_list:
    for model in model_lists:
        # Call the function
        mae, mse, r2 = wb_getmodels('precipitation',model,country)
        # Append results to DataFrame
        results_df = results_df.append({'Country': country, 'Data': 'Avg prec.','Model': model, 'MAE': mae, 'MSE': mse, 'R2': r2}, ignore_index=True)

[LightGBM] [Warning] Auto-choosing col-wise multi-threading, the overhead of testing was 0.000096 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 85
[LightGBM] [Info] Number of data points in the train set: 639, number of used features: 2
[LightGBM] [Info] Start training from score 55.388623
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, 

[LightGBM] [Warning] Auto-choosing col-wise multi-threading, the overhead of testing was 0.000118 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 85
[LightGBM] [Info] Number of data points in the train set: 639, number of used features: 2
[LightGBM] [Info] Start training from score 78.994147
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, 

Now, we will display the results:

In [ ]:
# Display the results
grouped_results = results_df.groupby('Country')

# Iterate over groups and print results
for country, country_results in grouped_results:
    print(f"Results for {country}:")
    print(country_results)
    print("\n")

Results for ALB:
    Country             Data          Model        MAE          MSE        R2
48      ALB  Avg surf. temp.            OLS   5.834023    46.180265 -0.043979
49      ALB  Avg surf. temp.  Decision Tree   1.432488     3.488849  0.919299
50      ALB  Avg surf. temp.            SVR   5.980983    45.829142 -0.013341
51      ALB  Avg surf. temp.       AdaBoost   1.392708     3.105787  0.936363
52      ALB  Avg surf. temp.       LightGBM   1.129001     2.179712  0.953569
53      ALB  Avg surf. temp.        XGBoost   1.338977     3.056766  0.934591
180     ALB        Avg prec.            OLS  55.922541  4481.008895 -0.017980
181     ALB        Avg prec.  Decision Tree  55.384038  5632.137803 -0.186159
182     ALB        Avg prec.            SVR  55.291405  5076.702195 -0.026049
183     ALB        Avg prec.       AdaBoost  54.745840  4478.816815  0.071668
184     ALB        Avg prec.       LightGBM  40.908079  2920.915100  0.267709
185     ALB        Avg prec.        XGBoost  50

LightGBM: Achieves the best performance with the lowest MAE, MSE, and highest R-squared. \
AdaBoost: Follows closely behind LightGBM with good performance. \
XGBoost: Performs well with a lower R-squared than LightGBM and AdaBoost. \
Decision Tree: Performs reasonably well. \
Linear Regression : Performs bad. \
SVR: Has the worst performance among the listed models.

| Country                | Country Code | Best Model  4  Precipitation | Best Model 4 Surf.Temp |
|------------------------|--------------|------------------------------|------------------------|
| Spain                  | ESP          |         AdaBoost             |       LightGBM         |
| France                 | FRA          |         AdaBoost             |       LightGBM         |
| Monaco                 | MCO          |         LightGBM             |       LightGBM         |
| Italy                  | ITA          |         AdaBoost             |       LightGBM         |
| Slovenia               | SVN          |         AdaBoost             |       LightGBM         |
| Croatia                | HRV          |         AdaBoost             |       LightGBM         |
| Bosnia and Herzegovina | BIH          |         LightGBM             |       LightGBM         |
| Montenegro             | MNE          |         LightGBM             |       LightGBM         |
| Albania                | ALB          |         LightGBM             |       LightGBM         |
| Greece                 | GRC          |         LightGBM             |       LightGBM         |
| Turkey                 | TUR          |         LightGBM             |        XGBoost         |
| Syria                  | SYR          |         LightGBM             |       LightGBM         |
| Lebanon                | LBN          |         AdaBoost             |        XGBoost         |
| Israel                 | ISR          |         AdaBoost             |       LightGBM         |
| Palestine              | PSE          |         AdaBoost             |       LightGBM         |
| Egypt                  | EGY          |          XGBoost             |       LightGBM         |
| Libya                  | LBY          |          XGBoost             |       LightGBM         |
| Tunisia                | TUN          |         LightGBM             |       LightGBM         |
| Algeria                | DZA          |         LightGBM             |       LightGBM         |
| Morocco                | MAR          |         AdaBoost             |       LightGBM         |
| Malta                  | MLT          |         LightGBM             |       LightGBM         |
| Cyprus                 | CYP          |         LightGBM             |       LightGBM         |


## Q4

Repeat what you did in Q3 for the precious metals data we used in HW2. Use the adjusted close column.

First, obtained the related data and rename the columns, then just by modifying the previous function, applied same steps to the commodity data.

In [ ]:
futures = yf.download(['SI=F','HG=F','PL=F','GC=F','PA=F'])
commodity = futures[['Adj Close']].dropna()
commodity.columns = ['gold','copper','palladium','platinum','silver']                                                               

[*********************100%%**********************]  5 of 5 completed


In [ ]:
def wb_getmodels2(comm_type,model_type):

    data = pd.DataFrame(commodity[comm_type]).reset_index()

    dates = pd.to_datetime(data['Date'])
    year = dates.dt.year
    month = dates.dt.month
    data2=pd.DataFrame({'year': year, 'month': month,'Commodity': data[comm_type]})
    X=data2[['year','month']]
    y = data2['Commodity']

    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25)

    # OLS
    if model_type == 'OLS':
        model = LinearRegression()
        model.fit(X_train, y_train)
    
    # Decision Tree Regression
    elif model_type == 'Decision Tree':
        model = DecisionTreeRegressor()
        model.fit(X_train, y_train)
    
    # SVM
    elif model_type == 'SVR':
        model = SVR()
        model.fit(X_train, y_train)
    
    # AdaBoost Regression
    elif model_type == 'AdaBoost':
        model = AdaBoostRegressor()
        model.fit(X_train, y_train)
    
    # LightGBM Regression
    elif model_type == 'LightGBM':
        model = LGBMRegressor()
        model.fit(X_train, y_train)
    
    # XGBoost Regression
    elif model_type == 'XGBoost':
        model = XGBRegressor()
        model.fit(X_train, y_train)
    else:
        print('Invalid model type')
    
    y_pred = model.predict(X_test)
    mae = mean_absolute_error(y_test, y_pred)
    mse = mean_squared_error(y_test, y_pred)
    r2 = r2_score(y_test, y_pred)
    
    return mae,mse,r2


In [ ]:
results2_df = pd.DataFrame(columns=['Commodity','Model', 'MAE', 'MSE', 'R2'])
for j in commodity.columns:
  for k in model_lists:
    mae, mse, r2 = wb_getmodels2(j,k)
    results2_df = results2_df.append({'Commodity': j, 'Model' : k , 'MAE': mae, 'MSE': mse, 'R2': r2}, ignore_index=True)

Commodity: gold--- model : OLS
Commodity: gold--- model : Decision Tree
Commodity: gold--- model : SVR
Commodity: gold--- model : AdaBoost
Commodity: gold--- model : LightGBM
[LightGBM] [Warning] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000764 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 39
[LightGBM] [Info] Number of data points in the train set: 3726, number of used features: 2
[LightGBM] [Info] Start training from score 1199.886312
Commodity: gold--- model : XGBoost
Commodity: copper--- model : OLS
Commodity: copper--- model : Decision Tree
Commodity: copper--- model : SVR
Commodity: copper--- model : AdaBoost
Commodity: copper--- model : LightGBM
[LightGBM] [Warning] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000414 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you

In [ ]:
# Display the results
grouped_results2 = results2_df.groupby('Commodity')

# Iterate over groups and print results
for commodity, commodity_results in grouped_results2:
    print(f"Results for {commodity}:")
    print(commodity_results)
    print("\n")

Results for copper:
   Commodity          Model       MAE       MSE        R2
6     copper            OLS  0.644711  0.589494  0.555858
7     copper  Decision Tree  0.056077  0.006868  0.994776
8     copper            SVR  0.898386  1.235628  0.055140
9     copper       AdaBoost  0.239205  0.102000  0.920848
10    copper       LightGBM  0.065516  0.009488  0.992435
11    copper        XGBoost  0.056133  0.007201  0.994573


Results for gold:
  Commodity          Model         MAE            MSE        R2
0      gold            OLS  184.639795   56723.511028  0.811612
1      gold  Decision Tree   17.006258     628.614076  0.997987
2      gold            SVR  440.991721  313547.600994 -0.012291
3      gold       AdaBoost   71.898316    8764.353874  0.970357
4      gold       LightGBM   19.913548     818.948628  0.997222
5      gold        XGBoost   17.042419     623.372165  0.997988


Results for palladium:
    Commodity          Model         MAE            MSE        R2
12  palladium  

### Results for Copper:
| Commodity | Model         | MAE      | MSE      | R2       |
|-----------|---------------|----------|----------|----------|
| Copper    | OLS           | 0.644711 | 0.589494 | 0.555858 |
| Copper    | Decision Tree | 0.056077 | 0.006868 | 0.994776 |
| Copper    | SVR           | 0.898386 | 1.235628 | 0.055140 |
| Copper    | AdaBoost      | 0.239205 | 0.102000 | 0.920848 |
| Copper    | LightGBM      | 0.065516 | 0.009488 | 0.992435 |
| Copper    | XGBoost       | 0.056133 | 0.007201 | 0.994573 |
| **Best Model** | **Decision Tree** | | | |

### Results for Gold:
| Commodity | Model         | MAE        | MSE           | R2       |
|-----------|---------------|------------|---------------|----------|
| Gold      | OLS           | 184.639795 | 56723.511028  | 0.811612 |
| Gold      | Decision Tree | 17.006258  | 628.614076    | 0.997987 |
| Gold      | SVR           | 440.991721 | 313547.600994 | -0.012291|
| Gold      | AdaBoost      | 71.898316  | 8764.353874   | 0.970357 |
| Gold      | LightGBM      | 19.913548  | 818.948628    | 0.997222 |
| Gold      | XGBoost       | 17.042419  | 623.372165    | 0.997988 |
| **Best Model** | **Decision Tree** | | | |

### Results for Palladium:
| Commodity  | Model         | MAE        | MSE           | R2       |
|------------|---------------|------------|---------------|----------|
| Palladium  | OLS           | 319.958301 | 170277.837830 | 0.581353 |
| Palladium  | Decision Tree | 32.441753  | 3386.010473   | 0.992332 |
| Palladium  | SVR           | 463.047104 | 448773.467395 | -0.073225|
| Palladium  | AdaBoost      | 127.542648 | 30097.256572  | 0.929889 |
| Palladium  | LightGBM      | 33.703387  | 3316.862031   | 0.991853 |
| Palladium  | XGBoost       | 31.971300  | 3439.288771   | 0.991985 |
| **Best Model** | **Decision Tree** | | | |

### Results for Platinum:
| Commodity | Model         | MAE        | MSE           | R2       |
|-----------|---------------|------------|---------------|----------|
| Platinum  | OLS           | 294.657248 | 126897.187837 | 0.029692 |
| Platinum  | Decision Tree | 23.510936  | 1151.467580   | 0.990812 |
| Platinum  | SVR           | 266.951161 | 131760.444931 | -0.071567|
| Platinum  | AdaBoost      | 82.278112  | 12543.428022  | 0.896526 |
| Platinum  | LightGBM      | 25.557000  | 1287.645185   | 0.989857 |
| Platinum  | XGBoost       | 23.399495  | 1265.944746   | 0.989951 |
| **Best Model** | **Decision Tree** | | | |

### Results for Silver:
| Commodity | Model         | MAE      | MSE      | R2       |
|-----------|---------------|----------|----------|----------|
| Silver    | OLS           | 5.047769 | 46.857902| 0.361755 |
| Silver    | Decision Tree | 0.499043 | 0.686071 | 0.990506 |
| Silver    | SVR           | 6.653344 | 74.315491| 0.011942 |
| Silver    | AdaBoost      | 1.735371 | 6.126845 | 0.920449 |
| Silver    | LightGBM      | 0.548417 | 0.712981 | 0.990281 |
| Silver    | XGBoost       | 0.525277 | 0.782943 | 0.990006 |
| **Best Model** | **Decision Tree** | | | |
